flang/lib/Lower/Allocatable.cpp

```diff
@@ -387,15 +387,20 @@ class MutablePropertyWriter {
     // Ignore lengths if already constant in the box type (this would trigger an
     // error in the embox).
     llvm::SmallVector<mlir::Value> cleanedLengths;
+    auto cleanedAddr = addr;
     if (auto charTy = box.getEleTy().dyn_cast<fir::CharacterType>()) {
+      // Cast address to box type so that both input and output type have
+      // unknown or constant lengths.
+      cleanedAddr =
+          builder.createConvert(loc, builder.getRefType(box.getBaseTy()), addr);
       if (charTy.getLen() == fir::CharacterType::unknownLen())
         cleanedLengths.append(lengths.begin(), lengths.end());
     } else if (box.isDerived()) {
       // TODO: derived type lengths clean-up
```

Comment:
Is this ok to let pass by silently for now?

Reply:
Probably not, thanks. I made it a `box.isDerivedWithLengthParameters()` + hard TODO.